### PR TITLE
Tensorflow use device jacobian products

### PR DESCRIPTION
**Pull request by albi3ro**
Follows on to PR #4654 .

Some quirks are still being worked out.

### CODE REVIEW

File: pennylane/interfaces/execution.py [hunk 1]
```diff
@@ -634,9 +634,13 @@ def post_processing(results):
     _grad_on_execution = False
 
     if config.use_device_jacobian_product and interface in jpc_interfaces:
+            if max_diff > 1:
+            raise NotImplementedError("no higher order derivatives with device derivatives")
```
Comment (albi3ro) [inline pennylane/interfaces/execution.py:638]:
```suggestion
                raise NotImplementedError("no higher order derivatives with device derivatives")
```

File: pennylane/interfaces/jax.py [hunk 2]
```diff
@@ -246,6 +246,7 @@ def _vjp_fwd(params, tapes, execute_fn, jpc):
 
 def _vjp_bwd(tapes, execute_fn, jpc, _, dy):
     """Perform the backward pass of a vjp calculation, returning the vjp."""
+    print(qml.math.unwrap(dy))
```
Comment (albi3ro) [inline pennylane/interfaces/jax.py:249]:
```suggestion
```

File: pennylane/devices/qubit/adjoint_jacobian.py
Comment (albi3ro) [inline pennylane/devices/qubit/adjoint_jacobian.py:1]: Here, we want to make sure that the output VJP has the same dtype as the input cotangents.  This fixes some of the float32/ float64 errors.

File: pennylane/gradients/general_shift_rules.py [hunk 3]
```diff
@@ -178,7 +179,7 @@ def _get_shift_rule(frequencies, shifts=None):
                 "may give unstable results for the parameter shift rules."
             )
 
-        coeffs = -2 * np.linalg.solve(sin_matrix.T, frequencies)
+        coeffs = -2 * linalg_solve(sin_matrix.T, frequencies)
```
Comment (albi3ro) [inline pennylane/gradients/general_shift_rules.py:182]: See Bugfix #4961 . This can sometimes cause bus errors with tensorflow.

File: pennylane/interfaces/tensorflow.py [hunk 4]
```diff
@@ -116,207 +165,110 @@ def _res_restructured(res, tapes):
...
-            if jacs:
-                # Jacobians were computed on execution
-                # No additional quantum evaluations needed; simply compute the VJPs directly.
-                vjps = _compute_vjp(dy, jacs, multi_measurements, has_partitioned_shots)
+            # reconstruct the nested structure of dy
+            nested_dy = _res_restructured(dy, tapes)
+
+            try:
+                vjps = jpc.compute_vjp(inner_tapes, nested_dy)
+            except AttributeError as e:
```
Comment (albi3ro) [inline pennylane/interfaces/tensorflow.py:252]: While not ideal, the alternative of using a `tf.py_function` causes an extra VJP call to occur. While caching could eliminate the biggest performance issues from having an extra call to `vjp_fn`, caching might be turned off. 

But it might an option to consider in the future.

File: pennylane/math/multi_dispatch.py [hunk 5]
```diff
@@ -729,7 +729,7 @@ def scatter_element_add(tensor, index, value, like=None):
     return np.scatter_element_add(tensor, index, value, like=like)
 
 
-def unwrap(values, max_depth=None):
+def unwrap(values, max_depth=None, keep_type=False):
```
Comment (albi3ro) [inline pennylane/math/multi_dispatch.py:732]: `keep_type` makes it so that the resulting number is a numpy array with the same precision. Will help us distinguish between float32 and float64.

File: pennylane/interfaces/execution.py [hunk 6]
```diff
@@ -707,6 +725,8 @@ def gradient_fn(internal_tapes):
                 return device.compute_derivatives(numpy_tapes, config)
 
     elif gradient_fn == "device":
+        # if max_diff > 1:
+        #    raise NotImplementedError("no higher order derivatives with device derivatives")
```
Comment (albi3ro) [inline pennylane/interfaces/execution.py:729]:
```suggestion
```

File: pennylane/interfaces/execution.py [hunk 7]
```diff
@@ -44,7 +44,17 @@
 
 device_type = Union[qml.Device, "qml.devices.Device"]
 
-jpc_interfaces = {"autograd", "numpy", "torch", "pytorch", "jax", "jax-python", "jax-jit"}
+jpc_interfaces = {
```
Comment (timmysilv) [inline pennylane/interfaces/execution.py:47]: is tensorflow-autograph the only one left?

File: pennylane/interfaces/tensorflow.py [hunk 8]
```diff
@@ -14,85 +14,134 @@
...
     structure of TF tensors
     """
-    if isinstance(x, dict) or isinstance(x, list) and all(isinstance(i, dict) for i in x):
+    if x is None or isinstance(x, dict):
         # qml.counts returns a dict (list of dicts when broadcasted), can't form a valid tensor
         return x
 
-    if isinstance(x, tuple):
+    if isinstance(x, (tuple, list)):
         return tuple(_to_tensors(x_) for x_ in x)
```
Comment (timmysilv) [inline pennylane/interfaces/tensorflow.py:134]: no risk converting lists to tuples here?

File: pennylane/interfaces/tensorflow.py [hunk 9]
```diff
@@ -116,207 +165,110 @@ def _res_restructured(res, tapes):
...
-            the maximum number of derivatives to support. Increasing this value allows
-            for higher order derivatives to be extracted, at the cost of additional
-            (classical) computational overhead during the backwards pass.
+        execute_fn (Callable[[Sequence[.QuantumTape]], ResultBatch]): a function that turns a batch of circuits into results
+        jpc (JacobianProductCalculator): a class that can compute the vector Jacobian product (VJP)
+            for the input tapes.
+
+    Keyword Args:
+        device=None: not used for tensorflow
+        differentiable = False: whether or not the custom gradient vjp needs to be
```
Comment (timmysilv) [inline pennylane/interfaces/tensorflow.py:179]: hopefully fixes rendering since it's the same as `device`
```suggestion
        differentiable=False: whether or not the custom gradient vjp needs to be
```

File: pennylane/interfaces/tensorflow.py [hunk 10]
```diff
@@ -116,207 +165,110 @@ def _res_restructured(res, tapes):
...
-            execute_fn
-            if not (logger.isEnabledFor(qml.logging.TRACE) and inspect.isfunction(execute_fn))
-            else "\n" + inspect.getsource(execute_fn) + "\n",
-            gradient_fn
-            if not (logger.isEnabledFor(qml.logging.TRACE) and inspect.isfunction(gradient_fn))
-            else "\n" + inspect.getsource(gradient_fn) + "\n",
-            gradient_kwargs,
-            _n,
-            max_diff,
-            "::L".join(str(i) for i in inspect.getouterframes(inspect.currentframe(), 2)[1][1:3]),
```
Comment (timmysilv) [inline pennylane/interfaces/tensorflow.py:189]: any reason you deleted the trace?

File: pennylane/interfaces/tensorflow.py [hunk 11]
```diff
@@ -116,207 +165,110 @@ def _res_restructured(res, tapes):
...
-    # pylint: disable=unused-argument
-
     parameters = []
-    params_unwrapped = []
+    numpy_params = []
 
-    # assumes all tapes have the same shot vector
-    has_partitioned_shots = tapes[0].shots.has_partitioned_shots
-
-    for i, tape in enumerate(tapes):
```
Comment (timmysilv) [inline pennylane/interfaces/tensorflow.py:200]: good thing we never used this `i`, because we re-define it _twice_ in this block 😅

File: pennylane/interfaces/tensorflow.py (hunk 4 above)
Comment (timmysilv) [reply pennylane/interfaces/tensorflow.py:252]: let's just add a story to track the decision in the future. side note: is the try-except the only reliable way to know when this error will happen ahead of time?

File: tests/interfaces/test_tensorflow_autograph_qnode_shot_vector.py [hunk 12]
```diff
@@ -55,6 +55,8 @@ def test_jac_single_measurement_param(
         """For one measurement and one param, the gradient is a float."""
         dev = qml.device(dev_name, wires=1, shots=shots)
 
+        print(dev_name, diff_method, gradient_kwargs, shots, num_copies, decorator, interface)
+
```
Comment (timmysilv) [inline tests/interfaces/test_tensorflow_autograph_qnode_shot_vector.py:59]:
```suggestion
```

**Review by timmysilv — COMMENTED** [review]
looks really good! I don't really mind the hiccup, but I'm not the most excited to have more experimental tensorflow (or any vendor, really) stuff referenced in our codebase. I agree it's an acceptable cost, but yeah..

thoughts on tracking the interface-specific kwargs in a unified spot? I'm not sure, but I do think they can get out of hand over time if we don't wrangle them down a little.

File: pennylane/interfaces/execution.py (hunk 7 above)
Comment (albi3ro) [reply pennylane/interfaces/execution.py:47]: Yep, though it will also be the worst. I'm starting to feel a bit more confident about figuring it out, but it might come down to whether or not it will be worth the effort.

File: pennylane/interfaces/tensorflow.py (hunk 8 above)
Comment (albi3ro) [reply pennylane/interfaces/tensorflow.py:134]: We are pretty haphazard about whether results are lists or tuples in a lot of places.  I personally prefer using a `tuple`, so out of the two, I'd prefer tuples. 

But we could also just use `type(x)` instead of `tuple`.

File: pennylane/interfaces/tensorflow.py (hunk 10 above)
Comment (albi3ro) [reply pennylane/interfaces/tensorflow.py:189]: I could probably add it back in.

File: pennylane/interfaces/tensorflow.py (hunk 4 above)
Comment (albi3ro) [reply pennylane/interfaces/tensorflow.py:252]: We could check if we are in eager execution and using a device vjp.

But then we start caring about the way we are calculating the vjp here, which this switchover was supposed to get rid of.

File: pennylane/interfaces/tensorflow.py (hunk 8 above)
Comment (timmysilv) [reply pennylane/interfaces/tensorflow.py:134]: doesn't seem to break anything so I'd say it's fine.

File: pennylane/devices/qubit/adjoint_jacobian.py [hunk 13]
```diff
@@ -399,9 +401,10 @@ def real_if_expval(val):
 
                 # Pad cotangent in with zeros for batch number with zero cotangents
                 cot_in = real_if_expval(np.sum(np.conj(bras) * ket_temp, axis=summing_axis))
+                cot_in = np.array(cot_in, dtype=dtype)
```
Comment (albi3ro) [inline pennylane/devices/qubit/adjoint_jacobian.py:404]:
```suggestion
```

File: pennylane/workflow/execution.py [hunk 14]
```diff
@@ -801,9 +819,16 @@ def device_gradient_fn(inner_tapes, **gradient_kwargs):
...
             execute_fn, gradient_fn, gradient_kwargs, cache_full_jacobian
         )
-        for _ in range(1, max_diff):
-            ml_boundary_execute = _get_ml_boundary_execute(interface, _grad_on_execution)
-            execute_fn = partial(ml_boundary_execute, execute_fn=execute_fn, jpc=jpc, device=device)
+        for i in range(1, max_diff):
+            ml_boundary_execute = _get_ml_boundary_execute(
+                interface, _grad_on_execution, differentiable=i > 1
+            )
+            execute_fn = partial(
```
Comment (rmoyard) [inline pennylane/workflow/execution.py:826]: Minor: Can you get `i > 1` out and assign it to a variable?

File: pennylane/workflow/interfaces/tensorflow.py [hunk 15]
```diff
@@ -116,207 +166,111 @@ def _res_restructured(res, tapes):
...
-                        str(i) for i in inspect.getouterframes(inspect.currentframe(), 2)[1][1:3]
-                    ),
-                )
+    def custom_gradient_execute(*parameters):  # pylint:disable=unused-argument
+        """An execution of tapes with VJP's registered with tensorflow.
 
-            # whether the tapes contain multiple measurements
-            multi_measurements = [len(tape.measurements) > 1 for tape in tapes]
+        Closure:
+            tapes (tuple(QuantumTape))
```
Comment (rmoyard) [inline pennylane/workflow/interfaces/tensorflow.py:226]: Minor: missing `: description`

**Review by rmoyard — COMMENTED** [review]
Thanks @albi3ro, it looks good to me! It is fine to restrict the vectorized case with `pfor=False`. Maybe I missed it but we need to raise an error for the case where `device_vjp + adjoint + state with unsupported float/complex`.

File: pennylane/workflow/interfaces/tensorflow.py (hunk 15 above)
Comment (albi3ro) [reply pennylane/workflow/interfaces/tensorflow.py]: added.

File: pennylane/workflow/execution.py (hunk 14 above)
Comment (albi3ro) [reply pennylane/workflow/execution.py:827]: assigned to `differentiable`.

**Review by rmoyard — APPROVED** [review]
Thanks, it looks good! :100:

File: pennylane/devices/default_qubit.py [hunk 16]
```diff
@@ -136,6 +138,18 @@ def adjoint_state_measurements(tape: QuantumTape) -> (Tuple[QuantumTape], Callab
         )
 
     params = tape.get_parameters()
+
+    if device_vjp:
+        for p in params:
+            if (
+                qml.math.requires_grad(p)
+                and qml.math.get_interface(p) == "tensorflow"
+                and "32" in qml.math.get_dtype_name(p)
```
Comment (timmysilv) [inline pennylane/devices/default_qubit.py:147]: dare I ask.. shouldn't we check for `complex64` here?

**Review by timmysilv — APPROVED** [review]
lol codefactor is unimpressed with `qml.execute`. lgtm!!

File: pennylane/devices/default_qubit.py (hunk 16 above)
Comment (albi3ro) [reply pennylane/devices/default_qubit.py:147]: Ah.... yes... yes we do :(

File: pennylane/workflow/interfaces/tensorflow_autograph.py [hunk 17]
```diff
@@ -25,14 +25,38 @@
...
+        dy_ = dy_ if has_partitioned_shots else (dy_,)
+        jac_ = jac_ if has_partitioned_shots else (jac_,)
+
+        shot_vjps = []
+        for d, j in zip(dy_, jac_):
+            # see xfail test: test_tensorflow_qnode_default_qubit_2.py:test_autograph_adjoint_multi_out
+            # And Issue #5078
+            # pragma: no cover
+            if multi:
+                shot_vjps.append(qml.gradients.compute_vjp_multi(d, j))
```
Comment (albi3ro) [inline pennylane/workflow/interfaces/tensorflow_autograph.py:49]:
```suggestion
            if multi: # pragma: no cover
                shot_vjps.append(qml.gradients.compute_vjp_multi(d, j))
```
